machineconfiguration/v1/types.go

```diff
@@ -745,3 +749,169 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineStateSpec `json:"spec"`
+	// +optional
+	Status MachineStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineState `json:"items"`
+}
+
+// MachineStateSpec describes the type of State we are managing
+type MachineStateSpec struct {
+	// Kind describes the type of MachineState
+	// +optional
+	Kind   MachineStateType   `json:"machineStateKind,omitempty"`
+	Config MachineStateConfig `json:"configuration"`
+}
+
+// MachineStateStatus holds the reported information on a particular MachineState
+type MachineStateStatus struct {
+	// Config Describes the Kind of MachineState
+	Config MachineStateConfig `json:"configuration"`
+	// MostRecentState MCO Object Mapped to a Progress Condition
+	MostRecentState map[string]ProgressionCondition `json:"progress"`
+	// Progression describes the 20 Most Recent MachineStates
+	Progression []ProgressionCondition `json:"allStates"`
+	// MostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"error"`
+	// Healthy reports if the state is Healthy Overall
+	Healthy bool `json:"healthy"`
```

Comment:
Does this need to be a structured property?  If a state was not healthy, I'd expect folks to want to know why, and ideally what they could do about it to make it likely to recover.  If this became a property that tried to convey "I'm not healthy, and here's why, and here's what you might want to do about it", it might turn into `MostRecentError`, although [I'm not clear on the `MostRecentError` semantics][1].  But if those are getting at the same thing, maybe this property could be dropped?  And if they are getting at different things, perhaps this property could turn into a `string` or a metav1 `Condition` or some such to convey more information in the "I'm not healthy" case?

[1]: https://github.com/openshift/api/pull/1596#discussion_r1337882517